### PR TITLE
GH-39515: [Python] Pass in type to `MapType.from_arrays`

### LINKED ISSUE #39515: Allow passing in the Type for the `MapArray.from_arrays`
### Describe the enhancement requested

Similar to `ListArray.from_arrays(self, offsets, values, DataType_type=None, MemoryPool_pool=None, mask=None)`

### Component(s)

Python

### PR DESCRIPTION
**Pull request by Fokko**
### Rationale for this change

For Iceberg we want to add metadata type the type (the field-id), therefore we need to pass in the type analog to what we do for `ListArray.from_arrays(self, offsets, values, DataType type=None, MemoryPool pool=None, mask=None)`.

### What changes are included in this PR?

Updated a keyword argument for the `type`, and make sure that the the static method to create the MapType is exposed from the cpp side.

### Are these changes tested?

I've added a simple test.

### Are there any user-facing changes?

* Closes: #39515

### CODE REVIEW

**Review by AlenkaF — APPROVED** [review]
Thank you for the contribution! LGTM 👍 with one small comment for the tests:

Could we add another simple test in case we are passing some other map type then default (keys' and items' type)? Also could we check for an error if the type is not `MapType` instance?

**Comment by Fokko** [comment]
Thanks for the quick review @AlenkaF 🙌 I've just added a couple more tests, let me know if this is what you're looking for 👍

File: python/pyarrow/tests/test_array.py [hunk 1]
```diff
@@ -1057,6 +1057,31 @@ def test_map_from_arrays():
...
+        pa.MapArray.from_arrays(offsets, keys, items, pa.string())
+
+    with pytest.raises(pa.ArrowTypeError, match='Mismatching map items type'):
+        pa.MapArray.from_arrays(offsets, keys, items, pa.map_(
+            keys.type,
+            # This is not allowed
+            pa.string()
+        ))
+
+    assert result.equals(expected)
```
Comment (ianmcook) [inline python/pyarrow/tests/test_array.py:1083]: Are these lines needed?

File: python/pyarrow/tests/test_array.py [hunk 2]
```diff
@@ -1057,6 +1057,31 @@ def test_map_from_arrays():
...
+        items.type
+    ))
+    assert result.equals(expected)
+
+    pa.MapArray.from_arrays(offsets, keys, items, pa.map_(
+        keys.type,
+        # Bigger than the original i4, which is okay
+        pa.int32()
+    ))
+    assert result.equals(expected)
```
Comment (ianmcook) [inline python/pyarrow/tests/test_array.py:1071]: Does this assertion serve a purpose?

Comment (AlenkaF) [reply python/pyarrow/tests/test_array.py:1071]: Yeah, I think `result = ...` is missing in line 1066.

And for the type-double-check, `pa.int32` and numpy `i4` should be same?
If we were to use `pa.int64` the construction of a MapArray should not fail, but the assertion would, right?

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python -g wheel

File: python/pyarrow/array.pxi [hunk 3]
```diff
@@ -2467,7 +2467,7 @@ cdef class MapArray(ListArray):
     """
 
     @staticmethod
-    def from_arrays(offsets, keys, items, MemoryPool pool=None):
+    def from_arrays(offsets, keys, items, DataType type=None, MemoryPool pool=None):
```
Comment (raulcd) [inline python/pyarrow/array.pxi:2470]: is there a reason to use `type` as a variable name? `type` is [a built-in function in Python](https://docs.python.org/3/library/functions.html#type) and unless necessary could we call it `datatype` or `data_type`?

Comment (AlenkaF) [reply python/pyarrow/array.pxi:2470]: Agree but at the same time, we seem to be using `type` elsewhere also. For example:
https://github.com/apache/arrow/blob/72ed58449ea71aab1343d9adce19f177f20705cf/python/pyarrow/array.pxi#L2139C15-L2145

Comment (raulcd) [reply python/pyarrow/array.pxi:2470]: oh! I see, I did a search and didn't found other occurrences. Thanks @AlenkaF for pointing that out

File: python/pyarrow/tests/test_array.py (hunk 2 above)
Comment (Fokko) [reply python/pyarrow/tests/test_array.py]: Thanks! I'm not too familiar with the `numpy` types, thanks for pointing that out!

**Review by ianmcook — APPROVED** [review]

**Comment by AlenkaF** [comment]
@github-actions crossbow submit -g python -g wheel

**Comment by Fokko** [comment]
Thanks @AlenkaF, @ianmcook and @raulcd for the quick review and merge 🙏

**Pull request merged**